prov/netdir/netdir_buf.h

```diff
@@ -0,0 +1,257 @@
+* General Public License (GPL) Version 2, available from the file
+* COPYING in the main directory of this source tree, or the
+* BSD license below:
+*
+*     Redistribution and use in source and binary forms, with or
+*     without modification, are permitted provided that the following
+*     conditions are met:
+*
+*      - Redistributions of source code must retain the above
+*        copyright notice, this list of conditions and the following
+*        disclaimer.
+*
+*      - Redistributions in binary form must reproduce the above
+*        copyright notice, this list of conditions and the following
+*        disclaimer in the documentation and/or other materials
+*        provided with the distribution.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+* NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+* BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+* ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+* CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+* SOFTWARE.
+*/
+
+#ifndef _FI_NETDIR_BUF_H_
+#define _FI_NETDIR_BUF_H_
+
+#include <windows.h>
+
+#include "rdma/fabric.h"
+
+#include "fi.h"
+#include "fi_osd.h"
+
+#include "netdir.h"
+#include "netdir_assert.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+	/* Implementation of lock-free thread safe allocator of fixed size data */
+
+	/*
+	Lock free allocator is implemented using atomics add/compare-and-swap/sub.
+	There is three levels of data:
+	footer - high level structure, holds links to first chunk, first non-used
+	   |     element, number of allocated elelements and number of used elements.
+	   |
+	   +-chunk - one-directional linked list where stored items
+	       |
+	       +-item - one directional linked list where stored user's data
+
+	chunk elements are linked to be freed only, nothing else used linked nature of
+	  chunk.
+	item elemens may have two states: used/not-used. All not-used (free) elements
+	  are linked into one-directional linked list, pointer to first non-used
+	  element is located at footer.
+
+	when user allocates new element function buf_alloc looks for number of used
+	  elements and compares it with allocated number of elements: used atomic
+	  API InterlockedIncrement to increment counter by 1 and check if there
+	  is non-used items. In case if no more available elements exist
+	  (used > counter) - allocate new chunk.
+
+	All pointers are updated using atomic API compare-and-swap which guarantee
+	  thread safe of access to allocator. Only init/fini calls are not thread safe.
+
+	Call of ofi_nd_buf_init_*** API is not mandatory - default initialization
+	  by NULL is enough
+
+	Note: underscored (__ofi_nd_buf_****) functions are used to implement footer
+	  inside of another objects
+	*/
```

Comment:
This sounds generic.  Please provide a patch to the common code to add this there, with names reworked appropriately.  It looks like we just need support for some atomic operations.